### ISSUE TITLE
Fix stack overflow when using deadlock detection in Pool::acquire

### ISSUE THREAD

**Issue by madadam** [equalitie/ouisync#48]
Currently the deadlock detection in `Pool::acquire` is disabled because it was causing stack overflow in the `network::tests::failed_block_other_peer` test. We should find out why it happens and fix it.

**Comment by madadam** [comment]
This might possible have nothing to do with the deadlock detection. I ran the test with `rust-gdb` and this is the backtrace:

```
#0  0x0000555556a98811 in __rust_probestack ()
#1  0x0000555556954432 in sha2::sha512::x86::sha512_compress_x86_64_avx2 (state=0x510e527fade682d1, blocks=struct &[[u8; 128]](size=2270897969802886507) = {...})
    at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/sha2-0.9.9/src/sha512/x86.rs:29
#2  0x000055555695440c in sha2::sha512::x86::compress (state=0x7ffff7a1ee50, blocks=struct &[[u8; 128]](size=1) = {...})
    at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/sha2-0.9.9/src/sha512/x86.rs:21
#3  0x0000555556952631 in sha2::sha512::compress512 (state=0x7ffff7a1ee50, blocks=
    struct &[generic_array::GenericArray<u8, typenum::uint::UInt<typenum::uint::UInt<typenum::uint::UInt<typenum::uint::UInt<typenum::uint::UInt<typenum::uint::UInt<typenum::uint::UInt<typenum::uint::UInt<typenum::uint::UTerm, typenum::bit::B1>, typenum::bit::B0>, typenum::bit::B0>, typenum::bit::B0>, typenum::bit::B0>, typenum::bit::B0>, typenum::bit::B0>, typenum::bit::B0>>](size=1) = {...}) at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/sha2-0.9.9/src/sha512.rs:261
#4  0x000055555695240a in sha2::sha512::Engine512::finish::{{closure}} (d=0x7ffff7a1edd0) at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/sha2-0.9.9/src/sha512.rs:38
#5  0x0000555556953817 in block_buffer::BlockBuffer<BlockSize>::len128_padding_be (self=0x7ffff7a1edc8, data_len=768, f=...)
    at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/block-buffer-0.9.0/src/lib.rs:181
#6  0x00005555569523ae in sha2::sha512::Engine512::finish (self=0x7ffff7a1edb8) at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/sha2-0.9.9/src/sha512.rs:37
#7  0x0000555556952486 in <sha2::sha512::Sha512 as digest::fixed::FixedOutputDirty>::finalize_into_dirty (self=0x7ffff7a1edb8, out=0x7ffff7a1f010)
    at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/sha2-0.9.9/src/sha512.rs:76
#8  0x000055555694f301 in <D as digest::fixed::FixedOutput>::finalize_into (self=..., out=0x7ffff7a1f010)
    at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/digest-0.9.0/src/fixed.rs:63
#9  0x000055555694f4a4 in digest::fixed::FixedOutput::finalize_fixed (self=...) at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/digest-0.9.0/src/fixed.rs:24
#10 0x000055555694f2df in <D as digest::digest::Digest>::finalize (self=...) at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/digest-0.9.0/src/digest.rs:71
#11 0x000055555694d6ab in curve25519_dalek::scalar::Scalar::from_hash (hash=...) at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/curve25519-dalek-3.2.1/src/scalar.rs:643
#12 0x000055555694c76e in <ed25519_dalek::public::PublicKey as signature::verifier::Verifier<ed25519::Signature>>::verify (self=0x7ffff0019ce0, message=struct &[u8](size=32) = {...}, 
    signature=0x7ffff7a20168) at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/ed25519-dalek-1.0.1/src/public.rs:346
#13 0x0000555555df63bf in ouisync::crypto::sign::PublicKey::verify (self=0x7ffff0019ce0, msg=struct &[u8](size=32) = {...}, signature=0x7ffff7a20168) at lib/src/crypto/sign.rs:74
#14 0x0000555555dcb4bc in ouisync::index::proof::UntrustedProof::verify (self=..., repository_id=0x7ffff0019ce0) at lib/src/index/proof.rs:85
#15 0x0000555555ecfd28 in ouisync::index::Index::receive_root_node::{{closure}} () at lib/src/index/mod.rs:139
#16 0x0000555556276ab9 in <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll (self=..., cx=0x7ffff7b1cd50)
    at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91
#17 0x00005555562f13fa in ouisync::network::client::Client::handle_root_node::{{closure}} () at lib/src/network/client.rs:170
#18 0x000055555627d011 in <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll (self=..., cx=0x7ffff7b1cd50)
    at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91
#19 0x00005555562f09e1 in ouisync::network::client::Client::handle_response::{{closure}} () at lib/src/network/client.rs:136
#20 0x00005555562677e1 in <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll (self=..., cx=0x7ffff7b1cd50)
    at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91
#21 0x00005555562ef4f9 in ouisync::network::client::Client::run::{{closure}} () at lib/src/network/client.rs:77
#22 0x00005555562773e1 in <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll (self=..., cx=0x7ffff7b1cd50)
    at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91
#23 0x0000555555c21f08 in ouisync::network::tests::simulate_connection::{{closure}}::{{closure}} (cx=0x7ffff7b1cd50)
    at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.1/src/macros/select.rs:507
#24 0x0000555555c2ed58 in <tokio::future::poll_fn::PollFn<F> as core::future::future::Future>::poll (self=..., cx=0x7ffff7b1cd50)
    at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.1/src/future/poll_fn.rs:38
#25 0x0000555555c1862b in ouisync::network::tests::simulate_connection::{{closure}} () at lib/src/network/tests.rs:467
#26 0x000055555626335b in <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll (self=..., cx=0x7ffff7b1cd50)
    at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91
#27 0x0000555555c2296f in ouisync::network::tests::run_until::{{closure}}::{{closure}} (cx=0x7ffff7b1cd50)
    at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.1/src/macros/select.rs:507
#28 0x0000555555c2f5db in <tokio::future::poll_fn::PollFn<F> as core::future::future::Future>::poll (self=..., cx=0x7ffff7b1cd50)
    at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.1/src/future/poll_fn.rs:38
#29 0x0000555555c1c087 in ouisync::network::tests::run_until::{{closure}} () at lib/src/network/tests.rs:484
#30 0x000055555627982b in <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll (self=..., cx=0x7ffff7b1cd50)
    at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91
#31 0x0000555555cb2b4d in <futures_util::future::future::fuse::Fuse<Fut> as core::future::future::Future>::poll (self=..., cx=0x7ffff7b1cd50)
    at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.21/src/future/future/fuse.rs:86
#32 0x0000555555ca7e4a in <core::pin::Pin<P> as core::future::future::Future>::poll (self=..., cx=0x7ffff7b1cd50)
    at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/future.rs:124
#33 0x0000555555ca3072 in <&mut F as core::future::future::Future>::poll (self=..., cx=0x7ffff7b1cd50)
    at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/future.rs:112
#34 0x0000555555c8bf0b in futures_util::future::future::FutureExt::poll_unpin (self=0x7ffff7a97258, cx=0x7ffff7b1cd50)
    at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.21/src/future/future/mod.rs:562
#35 0x00005555562025c3 in <futures_util::future::select::Select<A,B> as core::future::future::Future>::poll (self=..., cx=0x7ffff7b1cd50)
    at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.21/src/future/select.rs:103
#36 0x0000555555c21310 in ouisync::network::tests::failed_block_other_peer::{{closure}} () at lib/src/network/tests.rs:293
#37 0x00005555562690fb in <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll (self=..., cx=0x7ffff7b1cd50)
    at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91
#38 0x0000555555ca711a in <core::pin::Pin<P> as core::future::future::Future>::poll (self=..., cx=0x7ffff7b1cd50)
    at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/future.rs:124
#39 0x00005555561618c0 in tokio::runtime::basic_scheduler::CoreGuard::block_on::{{closure}}::{{closure}}::{{closure}} ()
    at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.1/src/runtime/basic_scheduler.rs:508
#40 0x0000555555e80090 in tokio::coop::with_budget::{{closure}} (cell=0x7ffff7c1b568) at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.1/src/coop.rs:102
#41 0x000055555607a836 in std::thread::local::LocalKey<T>::try_with (self=0x555556ab62a0, f=...) at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/thread/local.rs:442
#42 0x000055555604524d in std::thread::local::LocalKey<T>::with (self=0x555556ab62a0, f=...) at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/thread/local.rs:418
#43 0x0000555556154d4c in tokio::coop::with_budget (budget=..., f=...) at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.1/src/coop.rs:95
#44 tokio::coop::budget (f=...) at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.1/src/coop.rs:72
#45 tokio::runtime::basic_scheduler::CoreGuard::block_on::{{closure}}::{{closure}} ()
    at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.1/src/runtime/basic_scheduler.rs:508
#46 0x00005555560d5982 in tokio::runtime::basic_scheduler::Context::enter (self=0x7ffff7b1cfc8, core=0x7ffff000d4d0, f=...)
    at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.1/src/runtime/basic_scheduler.rs:362
#47 0x000055555611f08e in tokio::runtime::basic_scheduler::CoreGuard::block_on::{{closure}} (core=0x7ffff000d4d0, context=0x7ffff7b1cfc8)
    at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.1/src/runtime/basic_scheduler.rs:507
#48 0x0000555556111aab in tokio::runtime::basic_scheduler::CoreGuard::enter::{{closure}} ()
    at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.1/src/runtime/basic_scheduler.rs:565
#49 0x000055555636ab77 in tokio::macros::scoped_tls::ScopedKey<T>::set (self=0x555556ad4f58 <tokio::runtime::basic_scheduler::CURRENT>, t=0x7ffff7b1cfc8, f=...)
    at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.1/src/macros/scoped_tls.rs:61
#50 0x000055555610d9b7 in tokio::runtime::basic_scheduler::CoreGuard::enter (self=..., f=...)
    at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.1/src/runtime/basic_scheduler.rs:565
#51 0x0000555556113d20 in tokio::runtime::basic_scheduler::CoreGuard::block_on (self=..., future=...)
    at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.1/src/runtime/basic_scheduler.rs:498
#52 0x0000555556099e3f in tokio::runtime::basic_scheduler::BasicScheduler::block_on (self=0x7ffff7bf0180, future=...)
    at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.1/src/runtime/basic_scheduler.rs:174
#53 0x0000555556217a71 in tokio::runtime::Runtime::block_on (self=0x7ffff7bf0178, 
    future=<error reading variable: value of type `core::future::from_generator::GenFuture<ouisync::network::tests::failed_block_other_peer::{async_block_env#0}>' requires 172672 bytes, which is more than max-value-size>) at /home/adam/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.1/src/runtime/mod.rs:480
#54 0x0000555555ed8a49 in ouisync::network::tests::failed_block_other_peer () at lib/src/network/tests.rs:308
#55 0x0000555555c1fd2a in ouisync::network::tests::failed_block_other_peer::{{closure}} () at lib/src/network/tests.rs:242
#56 0x0000555555f016be in core::ops::function::FnOnce::call_once () at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/ops/function.rs:227
#57 0x00005555563bee33 in core::ops::function::FnOnce::call_once () at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/ops/function.rs:227
#58 test::__rust_begin_short_backtrace () at library/test/src/lib.rs:574
#59 0x00005555563bdb99 in <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once () at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/alloc/src/boxed.rs:1861
#60 <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once ()
    at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panic/unwind_safe.rs:271
#61 std::panicking::try::do_call () at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:492
#62 std::panicking::try () at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:456
#63 std::panic::catch_unwind () at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panic.rs:137
#64 test::run_test_in_process () at library/test/src/lib.rs:597
#65 test::run_test::run_test_inner::{{closure}} () at library/test/src/lib.rs:491
#66 0x00005555563880ce in test::run_test::run_test_inner::{{closure}} () at library/test/src/lib.rs:518
#67 std::sys_common::backtrace::__rust_begin_short_backtrace () at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:122
#68 0x000055555638d698 in std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}} () at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/thread/mod.rs:498
#69 <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once ()
--Type <RET> for more, q to quit, c to continue without paging--~
    at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panic/unwind_safe.rs:271
#70 std::panicking::try::do_call () at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:492
#71 std::panicking::try () at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:456
#72 std::panic::catch_unwind () at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panic.rs:137
#73 std::thread::Builder::spawn_unchecked_::{{closure}} () at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/thread/mod.rs:497
#74 core::ops::function::FnOnce::call_once{{vtable-shim}} () at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/ops/function.rs:227
#75 0x0000555556a73c93 in <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once () at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/alloc/src/boxed.rs:1861
#76 <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once () at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/alloc/src/boxed.rs:1861
#77 std::sys::unix::thread::Thread::new::thread_start () at library/std/src/sys/unix/thread.rs:108
#78 0x00007ffff7f6f609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#79 0x00007ffff7d3f133 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

So it seems the issue comes from the `sha2` library. There is a bug reported there which looks relevant: https://github.com/RustCrypto/hashes/issues/344, however, that bug was fixed in v0.9.9 which is the version we are using, yet we are still seeing this issue. :shrug:

**Comment by madadam** [comment]
https://github.com/equalitie/ouisync/commit/a08bebb816fd00ef317acac18b2060d0d6403307 re-enabled the deadlock detection, but only in non-test mode. This way we can still take advantage of it when debugging or non-automated testing while avoiding the crash in tests.
